### PR TITLE
Quicksort improvements

### PR DESCRIPTION
**Pull request by benjaminsavage**
Making some changes suggested by @martinthomson on #892. 

Currently tests do not pass... still debugging...

### CODE REVIEW

**Review by martinthomson — COMMENTED** [review]
Did you do the swapping on the last iteration?  This seems to miss that.

File: ipa-core/src/protocol/ipa_prf/quicksort.rs [hunk 1]
```diff
@@ -115,29 +114,24 @@ where
...
-                if *b {
-                    list.swap(i, j + range.start + 1);
+        let mut comp_it = comp.into_iter();
+        for mut range in ranges_to_sort.into_iter().filter(|r| r.len() >= 2) {
+            let pivot_index = range.next().unwrap();
+            let mut i = pivot_index + 1;
+            while let Some(n) = range.next() {
+                let comparison = comp_it.next().unwrap();
+                if comparison {
+                    list.swap(i, n);
```
Comment (benjaminsavage) [inline ipa-core/src/protocol/ipa_prf/quicksort.rs:124]: I like this a lot better. What do you think @martinthomson ?

**Review by martinthomson — APPROVED** [review]

**Comment by martinthomson** [comment]
That looks like a pretty substantial improvement in clarity (and maybe even efficiency).  Thanks Ben.

File: ipa-core/src/protocol/ipa_prf/quicksort.rs [hunk 2]
```diff
@@ -46,7 +46,9 @@ pub(crate) enum Step {
 /// It terminates once the stack is empty.
 /// # Errors
 /// Will propagate errors from transport and a few typecasts
-#[tracing::instrument(name = "quicksort_ranges_by_key_insecure", skip_all)]
```
Comment (richajaindce) [inline ipa-core/src/protocol/ipa_prf/quicksort.rs:49]: I think u accidentally deleted tracing logs. Can u bring them back?

**Comment by akoshelev** [comment]
I think this broke clippy and some unit tests

**Pull request merged**